### PR TITLE
win-wasapi: Use provided timestamp for process capture and handle flags

### PR DESCRIPTION
**Pull request by derrod**
### Description

Changes process audio capture to use the OS-provided timestamp instead of calculating our own. In tests this seems to resolve audio issues reported by various users (also see: https://github.com/stream-labs/obs-studio/pull/543).

Additionally adds some logging and handling for flags that may be set by `IAudioClient::GetBuffer()`:
- `AUDCLNT_BUFFERFLAGS_TIMESTAMP_ERROR` is treated as failure, buffer is free'd and capture reset
- `AUDCLNT_BUFFERFLAGS_DATA_DISCONTINUITY` is logged to debug, but ignored as it probably won't cause issues with libobs's audio data processing
- `AUDCLNT_BUFFERFLAGS_SILENT` now replaces the buffer with all zeroes to avoid processing potentially stale or garbage data (also logged to debug as it can get very spammy)

### Motivation and Context

Hopefully helps with #8064 (not closing that one quite yet).

### How Has This Been Tested?

A user with extremely reliable reproduction of the audio issue reported no problems during a test recording.

### Types of changes

- Bug fix (non-breaking change which fixes an issue)
- Tweak (non-breaking change to improve existing functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: plugins/win-wasapi/win-wasapi.cpp [hunk 1]
```diff
@@ -1138,17 +1138,42 @@ bool WASAPISource::ProcessCaptureData()
 			return false;
 		}
 
+		// ToDo figure out if we need to reset capture here
```
Comment (jpark37) [inline plugins/win-wasapi/win-wasapi.cpp:1141]: Is this comment leftover? Also the OBS single-line style tends to be `/* Comment */`

File: plugins/win-wasapi/win-wasapi.cpp [hunk 2]
```diff
@@ -1138,17 +1138,42 @@ bool WASAPISource::ProcessCaptureData()
...
+			 * sample size = 4 bytes (always float per InitFormat) */
+			uint32_t requiredBufSize = 4 * speakers * frames;
+			if (silence.size() < requiredBufSize)
+				silence.resize(requiredBufSize);
+
+			buffer = silence.data();
+		}
+
 		obs_source_audio data = {};
 		data.data[0] = (const uint8_t *)buffer;
```
Comment (jpark37) [inline plugins/win-wasapi/win-wasapi.cpp:1170]: Is this cast necessary? Maybe you can remove it while you're here.

File: plugins/win-wasapi/win-wasapi.cpp [hunk 3]
```diff
@@ -1138,17 +1138,42 @@ bool WASAPISource::ProcessCaptureData()
...
+		}
+
+		if (flags & AUDCLNT_BUFFERFLAGS_SILENT) {
+			blog(LOG_DEBUG, "[WASAPISource::ProcessCaptureData]"
+					" Silent flag is set.");
+
+			/* buffer size = frame size * number of frames
+			 * frame size = channels * sample size
+			 * sample size = 4 bytes (always float per InitFormat) */
+			uint32_t requiredBufSize = 4 * speakers * frames;
```
Comment (jpark37) [inline plugins/win-wasapi/win-wasapi.cpp:1162]: `speakers` needs to be `get_audio_channels(speakers)`

File: plugins/win-wasapi/win-wasapi.cpp (hunk 2 above)
Comment (derrod) [reply plugins/win-wasapi/win-wasapi.cpp:1170]: Both this and the case for `frames` do seem redundant, yeah.

File: plugins/win-wasapi/win-wasapi.cpp (hunk 1 above)
Comment (derrod) [reply plugins/win-wasapi/win-wasapi.cpp:1141]: Leftover, I'll remove it.

File: plugins/win-wasapi/win-wasapi.cpp (hunk 3 above)
Comment (derrod) [reply plugins/win-wasapi/win-wasapi.cpp:1162]: The enum is conveniently laid out such that this works anyway, but you're obviously right that this should go through that for correctness.

Comment (jpark37) [reply plugins/win-wasapi/win-wasapi.cpp]: No, it APPEARS conveniently laid out. Look at the end. :P

**Comment by jpark37** [comment]
Lookg good. Thanks!

**Comment by prgmitchell** [comment]
I have not experienced this issue myself to be able to test the fix, but I have had for people verify they are no longer having issues after running the artifacts from this PR. I can get logs tomorrow if needed, glad to see there is finally a fix here.

**Review by Lain-B — APPROVED** [review]

**Comment by SuslikV** [comment]
For me (VirtualBox environment, old 20H2 OS, isolated) it doesn't solves the "crackling" issue of the https://github.com/obsproject/obs-studio/issues/8064 . I tested against "obs-studio-windows-x64-c2ec1fb49.zip" building artifact. In the log OBS was labeled as "OBS 30.0.0-beta3-47-gc2ec1fb49 (64-bit, windows)". I'm not sure is it right artifact to test the process capture feature, but anyway. Only difference I noticed is few leaks in the log.
[2023-10-09 10-25-26.txt](https://github.com/obsproject/obs-studio/files/12843686/2023-10-09.10-25-26.txt)

**Pull request merged**